Adding generic C codegen libraries for data interop with C and credit-based messaging

krame505 opened the pull request:
This adds my GenCRepr and GenCMsg libraries. 

`GenCRepr` provides generic byte-aligned versions of `pack`/`unpack`, and can be used to generate a C library at elaboration time for packing and unpacking of the same data to and from an equivalent struct/tagged union representation.  

`GenCMsg` uses `GenCRepr` to generically implement messaging protocols with credit-based control to avoid ahead-of-line blocking.  Given a message interface, generics are used to synthesize an implementation Bluespec module and generate a corresponding C library.  

I don't think there is much to be gained from cleaning up the git history here, so if @quark17 wants to keep the history nice, I am fine with squashing and merging.

quark17 commented on Libraries/Makefile:
```diff
@@ -5,6 +5,8 @@ BUILD_ORDER = \
 	AMBA_TLM3 \
 	FPGA \
 	NonPipelinedMath \
+	GenCRepr \
```
Do you want to put these under a common directory hiearchy?

krame505 replied:
Done.

quark17 commented on testing/bsc.contrib/GenCMsg/CounterB2C.bs:
```diff
@@ -0,0 +1,68 @@
+import Vector
+import GetPut
+
+type Id = Bit 16 -- Try changing the size (will need to update in CounterIface.bsv)
```
What does this comment mean?  In `CounterIface.bsv`, there are only generic message definitions, of size 64.

krame505 replied:
This test case started off as more of a demo for people to experiment with, originally.  One thing to try is changing the size of the messages, which is easy and flexible with generics.  But the use of `import "BDPI"` means that the packed message sizes are also hardcoded in `CounterIface.bsv`, so any change that affects the packed message sizes will also need to be reflected there.

quark17 replied:
I understand all that. My concern is that this type is called Id and it is 16 bits wide. If I have to change something in another file to match, I would expect to find something in that other file that is the number 16 and/or has Id in the name. Neither of these exist in the other file. So either these files are already out of sync (should Id be 64 bits, to match the 64 found in the other file?) or else the comment might need to say more about what needs to change / how it corresponds to the other file.

krame505 replied:
The issue is that changing any of the types can affect the size of the packed messages as determined by the `GenCMsg` type class, which must also be hardcoded in the `import "BDPI"` functions.  I added some comments to explain.

quark17 commented on testing/bsc.contrib/contrib.tcl:
```diff
@@ -38,7 +38,7 @@ proc add_contrib_dirs_to_path { dirs } {
     }
 
-    set libdir "$contrib_inst/lib/Libraries"
+    set libdir "$contrib_inst/Libraries"
```
This change should be removed.  At the moment, the `inst` directory does contain a `lib` directory in the hiearchy.  Maybe we should reconsider the directory layout, but that's for a different PR.

krame505 replied:
Huh.  I'm not sure if there's something wrong with how I'm running the contrib tests, but the tests all failed to compile until I made this change:
```
Warning: Command line: (S0091)
  Removing a directory from the path specified with the -p flag, because it
  could not be opened.
  Directory:
    /home/krame505/hardware/bsc-contrib/lib/Libraries/GenCRepr
  System message:
    does not exist
```

krame505 replied:
Oh, I misinterpreted the README instructions as "the `BSCCONTRIBDIR` env variable should be the path to the bsc-contrib installation" not "the `inst` subdirectory of the `bsc-contrib` repo".

quark17 reviewed (commented):
I've reviewed whether this PR is OK to commit, but the libraries themselves I have I only skimmed.  Let me know if you want me to delve into them in more detail, to offer comments.

krame505 commented:
I believe I have addressed all of your comments.  Additionally, I moved a couple of general-purpose Python files for utilizing the generated C libraries to the library source folder - I hope that this is alright.

quark17 commented on Libraries/GenC/GenCMsg/Makefile:
```diff
@@ -0,0 +1,19 @@
+# Sets BUILDDIR, and BSC and BSCFLAGS if not set
+# and defines the install target
+include ../../common.mk
+
```
I notice that, in the `AMBA_TLM2/TLM/` directory, the `Makefile` installs specific files, that includes installing a PDF.  If you wanted to do something similiar here, to install the python script, that would be OK.

quark17 commented on Libraries/Makefile:
```diff
@@ -5,6 +5,7 @@ BUILD_ORDER = \
 	AMBA_TLM3 \
 	FPGA \
 	NonPipelinedMath \
+	GenC
```
If I had noticed, I would have suggested that you end this line with a slash.  That way, later commits won't need to edit this line to add a slash, and the blame for this line will remain as the original commit.
